types/react/index.d.ts

```diff
@@ -833,7 +833,8 @@ declare namespace React {
      * @version 16.8.0
      * @see https://reactjs.org/docs/hooks-reference.html#usestate
      */
-    function useState<S>(initialState: S | (() => S)): [S, Dispatch<SetStateAction<S>>];
+    // function useState<S extends Function>(lazyState: () => S): [S, Dispatch<SetStateAction<S>>];
+    function useState<S>(initialState: S extends Function ? () => S : (() => S) | S): [S, Dispatch<SetStateAction<S>>];
```

Comment:
`Function` is not allowed by the linter; and it seems that using a conditional type here is causing the compiler to infer stricter / const-like types for the input 😞

I haven't tested it, but try this:

```ts
function useState<T>(initializer:
  | () => T
  | Exclude<T, (...args: any[]) => any>
): [T, Dispatch<SetStateAction<T>>]
```

You also need to do the same change on `SetStateAction`.

Reply:
Right, it would work if the compiler wouldn't infer stricter. I could not find any resources on why this is happening.